scripts/migrations/2024_10_move_ean8_products_to_new_path.pl

```diff
@@ -0,0 +1,348 @@
+										print STDERR "nested dir with less than 13 digits: $dir/$dir2/$dir3/$dir4\n";
+										print $log "nested dir with less than 13 digits: $dir/$dir2/$dir3/$dir4\n";
+										$d++;
+										(($d % 1000) == 1) and print STDERR "$d products - $dir/$dir2/$dir3/$dir4\n";
+									}
+								}
+							}
+							closedir $dh4;
+						}
+
+					}
+					closedir $dh3;
+				}
+			}
+			closedir $dh2;
+
+		}
+		else {
+			# Product directories at the root, with a different number than 3 digits
+			if (-e "$data_root/products/$dir/product.sto") {
+				push @products, $dir;
+				$d++;
+				(($d % 1000) == 1) and print STDERR "$d products - $dir\n";
+			}
+		}
+	}
+	closedir $dh;
+}
+
+my $count = $#products;
+
+print STDERR "$count products to update\n";
+
+my $products_collection = get_products_collection();
+
+my $invalid = 0;
+my $moved = 0;
+my $not_moved = 0;
+my $same_path = 0;
+my $changed_code = 0;
+
+foreach my $old_path (@products) {
+
+	my $code = $old_path;
+	# remove / if any
+	$code =~ s/\///g;
+
+	my $new_code = normalize_code_zeroes($code);
+
+	my $product_id = product_id_for_owner(undef, $new_code);
+
+	my $path = new_product_path_from_id($product_id);
+
+	if ($path eq "invalid") {
+		$invalid++;
+		print STDERR "invalid path for code $code (old path: $old_path)\n";
+		print $log "invalid path for code $code (old path: $old_path)\n";
+	}
+	elsif ($path ne $old_path) {
+
+		if (    (!-e "$data_root/products/$path")
+			and (!-e "$www_root/images/products/$path"))
+		{
+			print STDERR "$path does not exist, moving $old_path\n";
+			print $log "$path does not exist, moving $old_path\n";
+			$moved++;
+
+			if ($move) {
+
+				my $prefix_path = $path;
+				$prefix_path =~ s/\/[^\/]+$//;    # remove the last subdir: we'll move it
+				ensure_dir_created_or_die("$data_root/products/$prefix_path");
+				ensure_dir_created_or_die("$www_root/images/products/$prefix_path");
+
+				if (    (!-e "$data_root/products/$path")
+					and (!-e "$www_root/images/products/$path"))
+				{
+					# File::Copy move() is intended to move files, not
+					# directories. It does work on directories if the
+					# source and target are on the same file system
+					# (in which case the directory is just renamed),
+					# but fails otherwise.
+					# An alternative is to use File::Copy::Recursive
+					# but then it will do a copy even if it is the same
+					# file system...
+					# Another option is to call the system mv command.
+
+					# In this script, we want to avoid creating copies as we are using zfs, so we use File::Copy move()
+
+					print STDERR ("moving product data $data_root/products/$old_path to $data_root/products/$path\n");
+					print $log ("moving product data $data_root/products/$old_path to $data_root/products/$path\n");
+
+					if (not move("$data_root/products/$old_path", "$data_root/products/$path")) {
+						print STDERR ("could not move product data: $!\n");
+						print $log ("could not move product data: $!\n");
```

Comment:
Shan't we count that as an error, and do a move-- and errors++

Also is it a good idea to continue with image if we failed for product data ?

Reply:
Good point, fixing